### ISSUE TITLE
[Question] Attention window for causal self-attention

### ISSUE THREAD

**Issue by wellescastro** [HazyResearch/flash-attention#89]
Greetings to everyone involved with Flash Attention, great work!

I was wondering if it would be possible to have some sort of attention window for a causal self-attention layer given the current flash attention implementation (CUDA backend). That could be helpful for cases where we are handling extremely long sequences and looking only at the previous _n_ sequence elements might be enough. Considering a vanilla attention implementation, that implies that instead of having a causal mask considering all the previous elements with something like 
`torch.triu(torch.ones((seqlen, seqlen), dtype=torch.bool, device=device), diagonal=1)`
we would have something like this:
`~torch.tril(torch.ones((seqlen, seqlen), dtype=torch.bool), diagonal=0) & torch.triu(torch.ones((seqlen, seqlen), dtype=torch.bool, device=device), diagonal=-window_size+1)`

**Comment by tridao** [comment]
Great question on long sequences. We don't support sliding window attention but there's an implementation at https://github.com/idiap/fast-transformers (called local attention). You could try that out.

One crude way of doing sth like this is to break up the sequence (say of length 16k) into 8 chunks of size 2k each, then do standard (dense) attention on each of the 2k chunk. 
We also have a proof-of-concept implementation of block sparse FlashAttention.

**Comment by wellescastro** [comment]
Hi @tridao, I appreciate your help. I'll take a look at block sparse FlashAttention and the other suggestions. Thanks!
